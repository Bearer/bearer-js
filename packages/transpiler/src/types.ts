import { TInputDecoratorOptions, TOutputDecoratorOptions } from '@bearer/types/lib/input-output-decorators'
import * as ts from 'typescript'

import Metadata from './metadata'

export type ComponentMetadata = {
  classname: string
  isRoot: boolean
  initialTagName: string
  fileName: string
  finalTagName: string
  name?: string
  imports?: string[]
  inputs?: TComponentInputDefinition[]
  outputs?: TComponentOutputDefinition[]
}

export type TComponentInputDefinition = {
  name: string
  type: 'string' | 'number'
  default: string | number
}

type TBasicFormat = 'string' | 'number' | 'boolean' | 'object' | 'any'
type TUnknown = 'unspecified'
export type TOuputFormat = TUnknown | { type: TBasicFormat } | { [key: string]: TBasicFormat | TBasicFormat[] }

export type TComponentOutputDefinition = {
  name: string
  eventName: string
  payloadFormat: TOuputFormat
}

export type TransformerOptions = {
  verbose?: true
  metadata?: Metadata
}

export type FileTransformerOptions = TransformerOptions & {
  outDir: string
  srcDir?: string
}

export type SourceCodeTransformerOptions = TransformerOptions & {
  srcDirectory: string
  buildDirectory: string
}

export type SpecComponent = {
  classname: string
  isRoot: boolean
  initialTagName: string
  group: string
  label: string
  input: any
  output: any
}

export type RootComponent = SpecComponent & {
  finalTagName: string
}

export type CompileSpec = { components: SpecComponent[] }

export type InputMeta = TInputDecoratorOptions & {
  propDeclarationName: string
  typeIdentifier?: ts.TypeNode
  intializer?: ts.Expression
  loadMethodName: string
  intentMethodName: string
  watcherName: string
}

export type CreateFetcherMeta = {
  intentName: string
  intentMethodName: string
}

export type TCreateLoadResourceMethod = {
  propertyReferenceIdName: string
  typeIdentifier?: ts.TypeNode
  propDeclarationName: string
  intentMethodName: string
  intentReferenceIdKeyName: string
  loadMethodName: string
  intentArguments?: string[]
}

export type TCreateLoadDataCall = {
  loadMethodName: string
}

export type TAddAutoLoad = {
  propertyReferenceIdName: string
  autoLoad: boolean
  loadMethodName: string
}

export type OutputMeta = TOutputDecoratorOptions & {
  propDeclarationName: string
  propDeclarationNameRefId: string
  intentMethodName: string
  intentName: string
  loadMethodName: string
  initializer: ts.Expression
  typeIdentifier?: ts.TypeNode
  propertyReferenceIdName: string
  autoLoad?: true | false
}
