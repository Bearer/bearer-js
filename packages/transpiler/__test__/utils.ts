import * as ts from 'typescript'
export default (sourceFile: ts.SourceFile) => {
  let names: string[] = []
  traverseNode(sourceFile, names)
  function traverseNode(node: ts.Node, names: string[]) {
    if (node.kind === ts.SyntaxKind.Decorator) names.push(node.expression.expression.escapedText)
    ts.forEachChild(node, node => {
      traverseNode(node, names)
    })
  }
  return names
}
