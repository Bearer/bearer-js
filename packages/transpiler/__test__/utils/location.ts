import * as path from 'path'

export const FIXTURES = path.join(__dirname, '..', '__fixtures__')
export const BUILD = path.join(__dirname, '..', '.build/src')

export function UnitFixtureDirectory(dir: string) {
  return path.join(FIXTURES, 'unit', dir)
}

export function BuildUnitFixtureDirectory(dir: string) {
  return path.join(BUILD, 'unit', dir)
}
