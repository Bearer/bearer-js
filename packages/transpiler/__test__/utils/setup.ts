import * as fs from 'fs'
import * as globby from 'globby'
import * as path from 'path'

const buildFolder = path.join(__dirname, '..', '.build/src')

process.env.BEARER_SCENARIO_ID = 'SPONGE_BOB'

globby.sync(['**/*.tsx', '**/*.ts', '**/*.json'], { cwd: buildFolder }).forEach(file => {
  const filePath = path.join(buildFolder, file)
  if (fs.existsSync(filePath)) {
    fs.unlinkSync(filePath)
  }
})
