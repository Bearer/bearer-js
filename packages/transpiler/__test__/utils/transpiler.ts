import Transpiler, { TranpilerOptions } from '../../src/index'

export function TranspilerFactory(options: Partial<TranpilerOptions>): Transpiler {
  const defaults: TranpilerOptions = {
    watchFiles: false,
    buildFolder: '../../../.build/'
  }

  return new Transpiler({
    ...defaults,
    ...options
  })
}
