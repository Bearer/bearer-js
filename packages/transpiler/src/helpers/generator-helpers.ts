import * as ts from 'typescript'

import { createOrUpdateComponentDidLoad } from '../../src/transformers/bearer'

import { Decorators } from './../constants'
import { CreateFetcherMeta, TAddAutoLoad, TCreateLoadDataCall, TCreateLoadResourceMethod } from './../types'
import { idName } from './name-helpers'

export function createFetcher(meta: CreateFetcherMeta) {
  return ts.createProperty(
    [
      ts.createDecorator(
        ts.createCall(ts.createIdentifier(Decorators.Intent), undefined, [ts.createLiteral(meta.intentName)])
      )
    ],
    undefined,
    meta.intentMethodName,
    undefined,
    undefined,
    undefined
  )
}

function propertyReferenceIdNames(meta: TCreateLoadResourceMethod, metaCollection: TCreateLoadResourceMethod[]) {
  return meta.intentArguments.map(name => {
    const metaInfo = metaCollection.find(meta => meta.propDeclarationName === name)
    if (metaInfo) {
      return ts.createPropertyAssignment(
        idName(name),
        ts.createPropertyAccess(ts.createThis(), metaInfo.propertyReferenceIdName)
      )
    }
    return ts.createPropertyAssignment(name, ts.createPropertyAccess(ts.createThis(), name))
  })
}

export function createLoadResourceMethod(meta: TCreateLoadResourceMethod, metaCollection: TCreateLoadResourceMethod[]) {
  const intentCall = ts.createCall(ts.createPropertyAccess(ts.createThis(), meta.intentMethodName), undefined, [
    ts.createObjectLiteral([
      ts.createPropertyAssignment(
        meta.intentReferenceIdKeyName,
        ts.createPropertyAccess(ts.createThis(), meta.propertyReferenceIdName)
      ),
      ...propertyReferenceIdNames(meta, metaCollection)
    ])
  ])
  const udapteState = ts.createArrowFunction(
    undefined,
    undefined,
    [
      ts.createParameter(
        undefined,
        undefined,
        undefined,
        ts.createObjectBindingPattern([ts.createBindingElement(undefined, undefined, 'data')]),
        undefined,
        ts.createTypeLiteralNode([
          ts.createPropertySignature(undefined, 'data', undefined, meta.typeIdentifier, undefined)
        ]),
        undefined
      )
    ],
    undefined,
    undefined,
    ts.createBlock(
      [
        ts.createIf(
          ts.createIdentifier('data'),
          ts.createBlock([
            ts.createStatement(
              ts.createBinary(
                ts.createPropertyAccess(ts.createThis(), meta.propDeclarationName),
                ts.SyntaxKind.EqualsToken,
                ts.createIdentifier('data')
              )
            )
          ])
        )
      ],
      true
    )
  )
  const promiseHandler = ts.createCall(ts.createPropertyAccess(intentCall, 'then'), undefined, [udapteState])
  return ts.createProperty(
    undefined,
    undefined,
    meta.loadMethodName,
    undefined,
    undefined,
    ts.createArrowFunction(
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      ts.createBlock([ts.createStatement(promiseHandler)], true)
    )
  )
}

export function createLoadDataCall(meta: TCreateLoadDataCall) {
  return ts.createStatement(
    ts.createCall(ts.createPropertyAccess(ts.createThis(), meta.loadMethodName), undefined, undefined)
  )
}

export function addAutoLoad(tsClass: ts.ClassDeclaration, meta: TAddAutoLoad): ts.ClassDeclaration {
  if (meta.autoLoad) {
    return createOrUpdateComponentDidLoad(tsClass, block =>
      ts.updateBlock(block, [
        ...block.statements,
        ts.createIf(
          ts.createPropertyAccess(ts.createThis(), meta.propertyReferenceIdName),
          ts.createBlock([createLoadDataCall(meta)])
        )
      ])
    )
  }
  return tsClass
}
