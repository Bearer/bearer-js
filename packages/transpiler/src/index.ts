import * as ts from 'typescript'
import * as fs from 'fs-extra'
import * as path from 'path'
import { getSourceCode } from './utils'
import ReplaceIntentDecorators from './transformers/replace-intent-decorator'
import BearerScenarioIdInjector from './transformers/scenario-id-accessor-injector'
import PropInjector from './transformers/prop-injector'
import PropBearerContextInjector from './transformers/prop-bearer-context-injector'
import PropImporter from './transformers/prop-importer'
import BearerStateInjector from './transformers/bearer-state-injector'
import BearerReferenceIdInjector from './transformers/reference-id-injector'
import RootComponentTransformer from './transformers/root-component-transformer'
import NavigatorScreenTransformer from './transformers/navigator-screen-transformer'
import ImportsImporter from './transformers/imports-transformer'
import { Metadata, SourceCodeTransformerOptions } from './types'
import generateMetadataFile from './transformers/generate-metadata-file'
import GatherMetadata from './transformers/gather-metadata'

export type TranpilerOptions = {
  ROOT_DIRECTORY?: string
  watchFiles?: boolean
  buildFolder?: string
  srcFolder?: string
  verbose?: boolean
}

export default class Transpiler {
  private service: ts.LanguageService
  private rootFileNames: string[] = []
  private subscribers: ts.MapLike<Array<() => void>> = {}

  private readonly ROOT_DIRECTORY
  private watchFiles = true
  private buildFolder = '.bearer/views'
  private srcFolder = 'views'
  private verbose = true
  private files: ts.MapLike<{ version: number }> = {}
  private metadata: Metadata = {
    components: []
  }

  private compilerOptionsts: ts.CompilerOptions = {
    module: ts.ModuleKind.CommonJS
  }

  constructor(options?: Partial<TranpilerOptions>) {
    Object.assign(this, options)
    this.ROOT_DIRECTORY = this.ROOT_DIRECTORY || process.cwd()
  }

  run() {
    this.refresh()

    if (!this.watchFiles) {
      this.stop()
    }
  }

  emitFiles = () => {
    // Now let's watch the files
    this.rootFileNames.forEach(fileName => {
      this.files[fileName] = { version: 0 }
      // First time around, emit all files
      this.emitFile(fileName)
      if (this.watchFiles) {
        // Add a watch on the file to handle next change
        fs.watchFile(fileName, { persistent: true, interval: 250 }, (curr, prev) => {
          // Check timestamp
          if (+curr.mtime <= +prev.mtime) {
            return
          }
          // Update the version to signal a change in the file
          this.files[fileName].version++
          // write the changes to disk
          this.emitFile(fileName)
        })
      }
    })
  }

  refresh() {
    this.clearWatchers()

    const config = ts.readConfigFile(path.join(this.BUILD_DIRECTORY, 'tsconfig.json'), ts.sys.readFile)

    if (config.error) {
      throw new Error(config.error.messageText as string)
    }

    const parsed = ts.parseJsonConfigFileContent(config, ts.sys, this.VIEWS_DIRECTORY)
    this.rootFileNames = parsed.fileNames
    if (!this.rootFileNames.length) {
      console.warn('[BEARER]', 'No file to transpile')
    }

    const servicesHost: ts.LanguageServiceHost = {
      getScriptFileNames: () => this.rootFileNames,
      getScriptVersion: fileName => this.files[fileName] && this.files[fileName].version.toString(),
      getScriptSnapshot: fileName => {
        if (!fs.existsSync(fileName)) {
          return undefined
        }

        return ts.ScriptSnapshot.fromString(fs.readFileSync(fileName).toString())
      },
      getCurrentDirectory: () => process.cwd(),
      getCompilationSettings: () => this.compilerOptionsts,
      getDefaultLibFileName: options => ts.getDefaultLibFilePath(options),
      getCustomTransformers: () => this.transformers,
      fileExists: ts.sys.fileExists,
      readFile: ts.sys.readFile,
      readDirectory: ts.sys.readDirectory
    }
    // Create the language service files
    this.service = ts.createLanguageService(servicesHost, ts.createDocumentRegistry())

    this.emitFiles()
  }

  stop() {
    this.clearWatchers()
    this.trigger('STOP')
  }

  clearWatchers(): void {
    this.rootFileNames.forEach(fileName => {
      fs.unwatchFile(fileName)
    })
  }

  on(event: string, callback: () => void) {
    this.subscribers[event] = this.subscribers[event] || []
    this.subscribers[event].push(callback)
  }

  private trigger = (eventName: string) => {
    ;(this.subscribers[eventName] || []).forEach(callback => {
      callback()
    })
  }

  get transformers(): ts.CustomTransformers {
    const verbose = true
    return {
      before: [
        GatherMetadata({ verbose, metadata: this.metadata }),
        RootComponentTransformer({ verbose, metadata: this.metadata }),
        BearerReferenceIdInjector({ verbose, metadata: this.metadata }),
        ReplaceIntentDecorators({ verbose, metadata: this.metadata }),
        BearerScenarioIdInjector({ verbose, metadata: this.metadata }),
        PropImporter({ verbose, metadata: this.metadata }),
        PropInjector({ verbose, metadata: this.metadata }),
        PropBearerContextInjector({ verbose, metadata: this.metadata }),
        BearerStateInjector({ verbose, metadata: this.metadata }),
        NavigatorScreenTransformer({ verbose, metadata: this.metadata }),
        ImportsImporter({ verbose, metadata: this.metadata }),
        dumpSourceCode({
          verbose: true,
          srcDirectory: this.VIEWS_DIRECTORY,
          buildDirectory: this.BUILD_SCR_DIRECTORY
        }),
        generateMetadataFile({ verbose: verbose, metadata: this.metadata, outDir: this.BUILD_SCR_DIRECTORY })
      ],
      after: []
    }
  }

  emitFile = (fileName: string) => {
    const output = this.service.getEmitOutput(fileName)

    if (!output.emitSkipped) {
      if (this.verbose) {
        console.log(`Emitting ${fileName}`)
      }
    } else {
      console.log(`Emitting ${fileName} failed`)
      this.logErrors(fileName)
    }
  }

  logErrors(fileName: string) {
    let allDiagnostics = this.service
      .getCompilerOptionsDiagnostics()
      .concat(this.service.getSyntacticDiagnostics(fileName))
      .concat(this.service.getSemanticDiagnostics(fileName))

    allDiagnostics.forEach(diagnostic => {
      let message = ts.flattenDiagnosticMessageText(diagnostic.messageText, '\n')
      if (diagnostic.file) {
        let { line, character } = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start!)
        console.log(`  Error ${diagnostic.file.fileName} (${line + 1},${character + 1}): ${message}`)
      } else {
        console.log(`  Error: ${message}`)
      }
    })
  }

  private get BUILD_DIRECTORY(): string {
    return path.join(this.ROOT_DIRECTORY, this.buildFolder)
  }

  private get BUILD_SCR_DIRECTORY(): string {
    return path.join(this.BUILD_DIRECTORY, 'src')
  }

  private get VIEWS_DIRECTORY(): string {
    return path.join(this.ROOT_DIRECTORY, this.srcFolder)
  }
}

function dumpSourceCode(
  { srcDirectory, buildDirectory }: SourceCodeTransformerOptions = { srcDirectory, buildDirectory }
): ts.TransformerFactory<ts.SourceFile> {
  return _transformContext => {
    return tsSourceFile => {
      let outPath = tsSourceFile.fileName
        .replace(srcDirectory, buildDirectory)
        .replace(/js$/, 'ts')
        .replace(/jsx$/, 'tsx')
      fs.ensureFileSync(outPath)
      fs.writeFileSync(outPath, getSourceCode(tsSourceFile))

      return tsSourceFile
    }
  }
}
