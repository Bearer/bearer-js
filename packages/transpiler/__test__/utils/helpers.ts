import * as fs from 'fs'
import * as globby from 'globby'
import * as path from 'path'
import * as ts from 'typescript'

import { TranpilerOptions } from '../../src/index'
import { UnitFixtureDirectory } from '../utils/location'

import { TranspilerFactory } from './transpiler'

const ROOT_DIRECTORY = path.join(__dirname, '..')
const buildSrcFolder = path.join(ROOT_DIRECTORY, '.build/src')

export function cleanBuildFolder() {
  const buildFolder = path.join(__dirname, '..', '.build/src')

  globby.sync(['**/*.tsx', '**/*.ts', '**/*.json'], { cwd: buildFolder }).forEach(file => {
    const filePath = path.join(buildFolder, file)
    if (fs.existsSync(filePath)) {
      fs.unlinkSync(filePath)
    }
  })
}

export function runTransformers(code: string, transformers: ts.TransformerFactory<ts.SourceFile>[]) {
  const sourceFile = ts.createSourceFile('tmp.ts', code, ts.ScriptTarget.Latest)
  const transformed = ts.transform(sourceFile, transformers)

  const printer = ts.createPrinter(
    {
      newLine: ts.NewLineKind.LineFeed
    },
    {
      onEmitNode: transformed.emitNodeWithNotification,
      substituteNode: transformed.substituteNode
    }
  )
  const result = printer.printBundle(ts.createBundle(transformed.transformed))
  transformed.dispose()

  return result
}

export function runUnitOn(name: string, transpilerOptions: Partial<TranpilerOptions> = {}) {
  const srcDirectory = UnitFixtureDirectory(name)

  beforeAll(() => {
    cleanBuildFolder()
    runTranspiler(`__fixtures__/unit/${name}`, transpilerOptions)
  })

  fs.readdirSync(srcDirectory).forEach(file => {
    describe(file.replace(/\.tsx?/, '').replace(/\-/g, ' '), () => {
      it('match snapshot', async done => {
        expect.assertions(1)
        const filePath = path.join(buildSrcFolder, file)
        fs.readFile(filePath, 'utf8', (_e, postContent) => {
          expect({
            postContent,
            file
          }).toMatchSnapshot()
          done()
        })
      })
    })
  })
}

export function runTranspiler(srcFolder: string, transpilerOptions: Partial<TranpilerOptions> = {}) {
  const transpiler = TranspilerFactory({
    buildFolder: '.build',
    srcFolder,
    ROOT_DIRECTORY,
    ...transpilerOptions
  })

  transpiler.run()
}
