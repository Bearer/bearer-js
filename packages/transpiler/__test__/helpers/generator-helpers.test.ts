import * as ts from 'typescript'

import { addAutoLoad, createFetcher, createLoadResourceMethod } from '../../src/helpers/generator-helpers'
import { CreateFetcherMeta } from '../../src/types'
import { runTransformers } from '../utils/helpers'

function dummyTransformer(fun: (meta: any, metaCollection?: any) => ts.Node, meta: any, metaCollection?: any) {
  return (context: ts.TransformationContext) => {
    const updateClass: ts.Visitor = (node: ts.Node) => {
      if (ts.isClassDeclaration(node)) {
        const prop = node as ts.ClassDeclaration

        return ts.updateClassDeclaration(
          prop,
          prop.decorators,
          prop.modifiers,
          prop.name,
          prop.typeParameters,
          prop.heritageClauses,
          [...prop.members, fun(meta, metaCollection) as ts.ClassElement]
        )
      }

      return ts.visitEachChild(node, updateClass, context)
    }

    return (sourceFile: ts.SourceFile): ts.SourceFile => {
      return ts.visitNode(sourceFile, updateClass)
    }
  }
}

describe('createFetcher', () => {
  it('generates stuff properly', () => {
    const code = `
     class C {}
    `

    const meta: CreateFetcherMeta = {
      intentName: 'intentName',
      intentMethodName: 'intentMethodName'
    }

    expect(runTransformers(code, [dummyTransformer(createFetcher, meta)])).toMatchSnapshot()
  })
})

describe('createLoadResourceMethod', () => {
  it('generates load resource method properly', () => {
    const code = `
     class C {}
    `

    const metaCollection = [
      {
        propertyReferenceIdName: 'aPropertyReferenceIdName',
        typeIdentifier: ts.createIdentifier('SomeType'),
        propDeclarationName: 'a',
        intentMethodName: 'intentMethodName',
        intentReferenceIdKeyName: 'aId',
        intentArguments: [],
        loadMethodName: 'loadMethodName'
      },
      {
        propertyReferenceIdName: 'bPropertyReferenceIdName',
        typeIdentifier: ts.createIdentifier('SomeType'),
        propDeclarationName: 'b',
        intentMethodName: 'intentMethodName',
        intentReferenceIdKeyName: 'bId',
        intentArguments: [],
        loadMethodName: 'loadMethodName'
      },
      {
        propertyReferenceIdName: 'propertyReferenceIdName',
        typeIdentifier: ts.createIdentifier('SomeType'),
        propDeclarationName: 'propDeclarationName',
        intentMethodName: 'intentMethodName',
        intentReferenceIdKeyName: 'intentReferenceIdKeyName',
        intentArguments: ['a', 'b'],
        loadMethodName: 'loadMethodName'
      }
    ]

    const meta = {
      propertyReferenceIdName: 'propertyReferenceIdName',
      typeIdentifier: ts.createIdentifier('SomeType'),
      propDeclarationName: 'propDeclarationName',
      intentMethodName: 'intentMethodName',
      intentReferenceIdKeyName: 'intentReferenceIdKeyName',
      intentArguments: ['a', 'b'],
      loadMethodName: 'loadMethodName'
    }
    expect(runTransformers(code, [dummyTransformer(createLoadResourceMethod, meta, metaCollection)])).toMatchSnapshot()
  })
})

describe('addAutoLoad', () => {
  it('generates class with autoloaded feature', () => {
    function dummyTransformer(fun: (node: ts.ClassDeclaration, meta: any) => ts.Node, meta: any) {
      return (context: ts.TransformationContext) => {
        const updateClass: ts.Visitor = (node: ts.Node) => {
          if (ts.isClassDeclaration(node)) {
            const prop = node as ts.ClassDeclaration
            return fun(prop, meta)
          }

          return ts.visitEachChild(node, updateClass, context)
        }

        return (sourceFile: ts.SourceFile): ts.SourceFile => {
          return ts.visitNode(sourceFile, updateClass)
        }
      }
    }
    const meta = {
      propertyReferenceIdName: 'propertyReferenceIdName',
      autoLoad: true,
      loadMethodName: 'loadMethodName'
    }
    expect(runTransformers('class C {}', [dummyTransformer(addAutoLoad, meta)])).toMatchSnapshot()
  })
})
