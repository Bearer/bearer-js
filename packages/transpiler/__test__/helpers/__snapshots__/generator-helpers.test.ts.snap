// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addAutoLoad generates class with autoloaded feature 1`] = `
"class C {
    componentDidLoad() {
        if (this.propertyReferenceIdName) {
            this.loadMethodName();
        }
    }
}
"
`;

exports[`createFetcher generates stuff properly 1`] = `
"class C {
    @Intent(\\"intentName\\")
    intentMethodName;
}
"
`;

exports[`createLoadResourceMethod generates load resource method properly 1`] = `
"class C {
    loadMethodName = () => {
        this.intentMethodName({ intentReferenceIdKeyName: this.propertyReferenceIdName, aId: this.aPropertyReferenceIdName, bId: this.bPropertyReferenceIdName }).then(({ data }: {
            data: SomeType;
        }) => {
            if (data) {
                this.propDeclarationName = data;
            }
        });
    };
}
"
`;
